feat(Collect) Attempt to redesign the Collector step

fpacifici opened the pull request:
The current CollectStep is fairly confusing. It is not only batching messages but it embeds the logic that executes application code when the batch is full. This logic is itself represented as a ProcessingStrategy, but the Collector is not connected to this logic through ProcessingStrategy chaining as usual. 

The idea here is to make the collect idea more modular. 
It introduces a `BatchStep` and an `UnbatchStep`. These processing strategies are simpler. The first accumulates messages in a batch and submit the batch to the next step, the second does the opposite. It takes a batch and submits individual messages.

The idea is that the CollectStep and the ParallelCollectStep could be simply built by chaining two steps.
`BatchStep` -> `RunTask`

This process though has a few limitations.
1) Summarized here: https://github.com/getsentry/arroyo/issues/125. Processing strategies, today, takes a Message object in the submit method. This requires the message object to have a partition and an offset. When the message object represents a batch, partition and offset makes no sense. So we need to decide what to do about this scenario. Either we preserve the interface as it is and give up on modularity (or find a different way to make pipelines modular). Or we change the interface.
2) `RunTask` only commits one message, It is not capable of committing multiple offsets for different partitions for a batch. But this is easily fixable.

`BatchStep` and `UnbatchStep` manage `RejectedMessage` exceptions from the following steps by preserving the state and retrying.

volokluev commented on arroyo/processing/strategies/collect.py:
```diff
@@ -21,7 +32,210 @@ class OffsetRange:
+        batch = self.__batch_builder.flush()
+        try:
+            if len(batch) > 0:
+                last_msg = batch.messages[-1]
```
this seems like something you can keep track of in the `MessageBatch`

fpacifici replied:
Not sure I follow. What would you keep track of in `MessageBatch` ?

volokluev replied:
the last message

fpacifici replied:
done

volokluev commented on arroyo/processing/strategies/collect.py:
```diff
@@ -21,7 +32,210 @@ class OffsetRange:
+        return f"<{type(self).__name__}: {len(self)} message{'s' if len(self) != 1 else ''}, open for {self.duration:0.2f} seconds>"
+
+
+class BatchBuilder(Generic[TPayload]):
```
You kind of have to read the code of this class to figure out why it exists. As far as I can tell its main purpose is to keep track of the highest offsets per partition of any message. Is it making the assumption in `append` that offsets are strictly increasing? Those assumptions (or any others I've missed) should be stated

fpacifici replied:
Behavior changed and comment added

volokluev reviewed (commented):
This would also benefit from a minimal example of how these are actually supposed to be used

lynnagara commented on arroyo/processing/strategies/collect.py:
```diff
@@ -21,7 +32,210 @@ class OffsetRange:
+
+    messages: Sequence[Message[TPayload]]
+    offsets: Mapping[Partition, OffsetRange]
+    duration: float
```
Do we actually need it? Maybe we can omit?

fpacifici replied:
removed

fpacifici commented on arroyo/processing/strategies/collect.py:
```diff
@@ -21,7 +32,210 @@ class OffsetRange:
+        self.__max_batch_time = max_batch_time
+        self.__max_batch_size = max_batch_size
+        self.__messages: MutableSequence[Message[TPayload]] = []
+        self.__offsets: MutableMapping[Partition, OffsetRange] = {}
```
Unfortunately defaultdict would not work here as there is no way to build a default value for a missing key. So defining a default factory would not be possible.
When the key is not present we add a new OffsetRange with the offset of the current message. 
When we update the existing key we only update some of the fields of the OffsetRange, the original offset stays in place.

fpacifici commented on arroyo/processing/strategies/collect.py:
```diff
@@ -21,7 +32,210 @@ class OffsetRange:
+        return f"<{type(self).__name__}: {len(self)} message{'s' if len(self) != 1 else ''}, open for {self.duration:0.2f} seconds>"
+
+
+class BatchBuilder(Generic[TPayload]):
```
I don't think I like the idea of following the `submit` and `poll` approach for the BatchBuilder. 
That would make the interface look like a ProcessingStrategy while it is not. That would be confusing and it is what happens today with the collector:
https://github.com/getsentry/arroyo/blob/main/arroyo/processing/strategies/collect.py#L24

It sends me off every time when I read the code that uses it without having the class implementation in available.

I think I will try to get rid of the builder entirely

fpacifici replied:
In the end I could not remove the builder entirely. 
I changed the interface to have only two methods:
```
append
build_if_ready
```

I removed the scenario that required the force

lynnagara replied:
You do not have to use the `submit` and `poll` names if they are too close to a strategy. My reasons for suggesting something else were:

- `append` normally implies appending items to a list to me. Here you have another builder class so feels a bit different from append to me
- The main reason for suggesting the poll-like interface is that I think it is unnecessary to reinitialise the Builder class over and over for each batch. Whether or not you call it poll doesn't matter. But I think you should just have a single instance of it, as it will make BatchStep simpler.

fpacifici commented on arroyo/processing/strategies/collect.py:
```diff
@@ -21,7 +32,210 @@ class OffsetRange:
+    offsets: Mapping[Partition, OffsetRange]
+    duration: float
+
+    def __len__(self) -> int:
+        return len(self.messages)
```
Adding __iter__ and __getitem__

fpacifici commented on arroyo/processing/strategies/collect.py:
```diff
@@ -83,6 +297,8 @@ def join(self, timeout: Optional[float] = None) -> None:
 
 class CollectStep(ProcessingStep[TPayload]):
     """
+    DEPRECATED: Use BatchStep instead and add the step_factory logic as a next step.
```
Thinking over this again, it is too early to deprecate this method.

fpacifici commented on arroyo/processing/strategies/collect.py:
```diff
@@ -21,7 +32,210 @@ class OffsetRange:
+
+    def __init__(
+        self,
+        max_batch_time: float,
```
A float timestamp means seconds. That should remove ambiguity. Anyway added `_sec`

fpacifici commented on arroyo/processing/strategies/collect.py:
```diff
@@ -83,6 +297,8 @@ def join(self, timeout: Optional[float] = None) -> None:
 
 class CollectStep(ProcessingStep[TPayload]):
     """
+    DEPRECATED: Use BatchStep instead and add the step_factory logic as a next step.
```
Thinking over this again, it is too early to deprecate this method.

fpacifici commented on arroyo/processing/strategies/collect.py:
```diff
@@ -83,6 +297,8 @@ def join(self, timeout: Optional[float] = None) -> None:
 
 class CollectStep(ProcessingStep[TPayload]):
     """
+    DEPRECATED: Use BatchStep instead and add the step_factory logic as a next step.
```
Thinking over this again, it is too early to deprecate this method.

fpacifici commented on arroyo/processing/strategies/collect.py:
```diff
@@ -21,7 +32,210 @@ class OffsetRange:
+        self.__batch_builder: Optional[BatchBuilder[TPayload]] = None
+        self.__closed = False
+
+    def __flush(self, robust: bool) -> None:
```
refactored. This does not exist anymore

fpacifici commented:
More changes to come

fpacifici commented on arroyo/processing/strategies/collect.py:
```diff
@@ -21,7 +32,210 @@ class OffsetRange:
+            raise MessageRejected
+
+        self.__batch_to_send.extend(message.payload.messages)
+        self.__flush()
```
fixed

fpacifici commented:
> I think you should rewrite https://github.com/getsentry/arroyo/blob/main/arroyo/processing/strategies/factory.py to use the new strategy. It may make it more obvious how the new BatchStep is going to replace Collect and ParallelCollect inside the arroyo library itself

@lynnagara , I'd say, not yet. There are quite a few corner cases to deal with with batch and unbatch. I'd rather try it first on a less widespread scenario. `ConsumerStrategyFactory` powers every single snuba consumer. 

Will add the unit test and the exmple

fpacifici commented:
End to end test added. 
Last missing thing is the example

fpacifici commented on arroyo/processing/strategies/batching.py:
```diff
@@ -262,3 +265,265 @@ def create_with_partitions(
+
+    def join(self, timeout: Optional[float] = None) -> None:
+        """
+        Terminates the strategy by joining the following step.
+        This method throws away the current batch.
```
In the previous implementation I was flushing the batch here. 
Though I noticed some inconsistency in how we manage the `join` in the code base. Some steps like the tickbuffer or the parallel consumer abandon the work they were doing, while the collector flushes it. 

I think we need to get decide how to do this and stick to it.
What do you think? Ideally I would prefer flushing on join but then I get stuck in how to deal with MessageRejected. The join would have to loop till the message gets accepted potentially hitting the deadline and preventing the join of the following step from happening.

fpacifici commented:
Example added

onewland commented on examples/transform_and_produce/README.md:
```diff
@@ -14,7 +16,26 @@ The process is fairly simple:
+- The consumer consumes messages from the `raw-topic`
+- This message is in the form `{"username": "<username>", "password": "<password>"}`
+- Messages are batched
+- The batch is submitted to the `TransformStrategy` which would run some indexing logic on the whole
+  batch.
+- The results are exploded into messages again.
+- The indexed messages are then submitted to the `ProduceAndCommitStep` which simply produces the given message to `hash-topic`
```
Does indexing/indexed have a standard definition in stream processing? This might be as clear as we need but a quick google search didn't immediately explain to me what "indexing logic" means

fpacifici replied:
Not really, that is just an example. We could replace indexing, with any other process we would do on batches, like storing fields, encode them into something, run a particle accelerator with that data, etc.

fpacifici commented on arroyo/processing/strategies/batching.py:
```diff
@@ -262,3 +265,265 @@ def create_with_partitions(
+        """
+        self.__closed = True
+
+    def join(self, timeout: Optional[float] = None) -> None:
```
Should we? This was my question here 
https://github.com/getsentry/arroyo/pull/127#discussion_r1020635450

Should we stick to a general policy or should this be a case by case issue? 
I think flushing here for batching would be a good idea.

lynnagara replied:
I think we should. Afaik all the other strategies always continue processing (up to the timeout passed if any). It is true that we don't usually pass a timeout to join, but still I see no justification not to apply the general policy in just this one place.

fpacifici replied:
Brought flushing during join back.

lynnagara commented on examples/transform_and_produce/README.md:
```diff
@@ -14,7 +16,26 @@ The process is fairly simple:
 - The `ProduceAndCommitStep` is also responsible for commiting the original offset back for the consumer
 
-### Usage
+## Consume -> Batch -> Transform -> Unbatch -> Produce
```
Part of my motivation for asking this is that I'm not really sure that Unbatch is generically useful and I'm wanting to better understand real world use cases for it.

fpacifici replied:
Not sure I follow this comment. This is a separate example, it is not a variation of the existing one. The only common element between the existing (simple) one and the batched one is the startup script.

lynnagara replied:
Ignore my comment, this is fine. I looked at the metrics indexer code and I have a good idea of the use case from that now.

lynnagara commented on arroyo/processing/strategies/batching.py:
```diff
@@ -262,3 +265,265 @@ def create_with_partitions(
+
+    def append(self, message: Message[TPayload]) -> None:
+        if message.partition in self.__offsets:
+            if self.__offsets[message.partition].hi >= message.next_offset:
+                raise OutOfOrderMessage(
+                    f"Received offset {message.next_offset}. Current watermark {self.__offsets[message.partition].hi}"
+                )
```
I would not bother with this. I think it's safe to assume messages are received in order. None of the other strategies do checks like this and I think it's unnecessary for everyone to be doing this.

fpacifici replied:
Removed

fpacifici commented on arroyo/types.py:
```diff
@@ -90,3 +90,12 @@ def __call__(
+
+    lo: int  # inclusive
+    hi: int  # exclusive
+    timestamp: datetime
```
This code does not exist anymore

fpacifici commented on arroyo/processing/strategies/transform.py:
```diff
@@ -43,7 +43,7 @@ class ChildProcessTerminated(RuntimeError):
 
 
-class TransformStep(ProcessingStep[TPayload]):
+class TransformStep(ProcessingStep[TPayload], Generic[TPayload, TTransformed]):
```
It did not. It was somehow needed in a previous version that relied on different data structure. Cannto remember anymore why.

fpacifici replied:
It was needed because this class depends on two generic parameters: `TPayload` and `TTransformed`. It inherits `TPayload` from the parent class, but it needs to tell the type checker that it has another parameter.

If I do not add it, this part breaks:
https://github.com/getsentry/arroyo/pull/127/files#diff-52195e990ade00fc5326c72366307080a8b5fc04d9ac242a26538679d02bffb1R443-R450

```
next_step: TransformStep[ValuesBatch[str], ValuesBatch[str]] = TransformStep(
```

would be invalid as the type checker expects only one parameter. Thus it should become 

```
next_step: TransformStep[ValuesBatch[str]] = TransformStep(
```

but then the definition of the transformation function would be invalid:

```
def transformer(msg: Message[ValuesBatch[str]],) -> ValuesBatch[str]:
```

Would have to be a 

```
Callable[[ValuesBatch[str]], Nothing]
```

as the code that uses the transformation step does not know about `TTransformed`

fpacifici commented:
@lynnagara  I think I adapted it to the changes introduced by https://github.com/getsentry/arroyo/pull/134 and addressed all the comments but the one on the example. I need clarifications on that one.
Now I guess it needs to be reviewed again as the whole implementation is quite different.

lynnagara commented on examples/transform_and_produce/README.md:
```diff
@@ -14,7 +16,26 @@ The process is fairly simple:
+- The batch is submitted to the `TransformStrategy` which would run some indexing logic on the whole
+  batch.
+- The results are exploded into messages again.
+- The indexed messages are then submitted to the `ProduceAndCommitStep` which simply produces the given message to `hash-topic`
+- The `ProduceAndCommitStep` is also responsible for commiting the original offset back for the consumer
```
Mind updating these lines now it's a separate `Produce` and `Commit` step?

fpacifici replied:
done

lynnagara commented on arroyo/processing/strategies/batching.py:
```diff
@@ -233,3 +243,234 @@ def __flush(self) -> None:
+        self.__init_time = time.time()
+
+    def append(self, value: BaseValue[TPayload]) -> None:
+        committables = value.committable
+        for partition, position in committables.items():
+            if (
+                partition not in self.__offsets
+                or self.__offsets[partition].offset <= position.offset
+            ):
+                self.__offsets[partition] = position
```
Messages are in order, this can be a lot simpler.
```suggestion
        self.__offsets.update(value.committable)
```

fpacifici replied:
done

lynnagara commented on arroyo/processing/strategies/__init__.py:
```diff
@@ -22,4 +23,7 @@
     "RunTaskInThreads",
+    "BatchStep",
+    "UnbatchStep",
+    "ValuesBatch",
```
I don't think it's needed for most people - let's not export ValuesBatch from here?

fpacifici replied:
done

lynnagara reviewed (commented):
Mostly looks good to me, a couple of small things inline
